formats/json/jvmMain/src/kotlinx/serialization/json/internal/JsonStringBuilder.kt

```diff
@@ -0,0 +1,90 @@
+package kotlinx.serialization.json.internal
+
+/*
+ * Optimized version of StringBuilder that is specific to JSON-encoding
+ */
+internal actual class JsonStringBuilder {
+    private var array = CharArrayPool.take()
+    private var size = 0
+
+    actual fun append(value: Long) {
+        ensureAdditionalCapacity(20) // Long length
+        append(value.toString())
+    }
+
+    actual fun append(ch: Char) {
+        ensureAdditionalCapacity(1)
+        array[size++] = ch
+    }
+
+    actual fun append(string: String) {
+        val length = string.length
+        ensureAdditionalCapacity(length)
+        string.toCharArray(array, size, 0, string.length)
+        size += length
+    }
+
+    actual fun appendQuoted(string: String) {
+        ensureAdditionalCapacity(string.length + 2)
+        val arr = array
+        var sz = size
+        arr[sz++] = '"'
+        val length = string.length
+        string.toCharArray(arr, sz, 0, length)
+        for (i in sz until sz + length) {
+            val ch = arr[i].toInt()
+            // Do we have unescaped symbols?
+            if (ch < ESCAPE_MARKERS.size && ESCAPE_MARKERS[ch] != 0.toChar()) {
+                // Go to slow path
+                return appendStringSlowPath(i - sz, i, string)
+            }
+        }
+        // Update the state
+        sz += length
+        arr[sz++] = '"'
+        size = sz
+    }
+
+    private fun appendStringSlowPath(firstEscapedChar: Int, currentSize: Int, string: String) {
+        var sz = currentSize
+        for (i in firstEscapedChar until string.length) {
+            val ch = string[i].toInt()
+            // Do we have unescaped symbols?
+            var marker: Char = 0.toChar()
+            if (ch < ESCAPE_MARKERS.size && ESCAPE_MARKERS[ch].also { marker = it } != 0.toChar()) {
+                if (marker != 0.toChar()) {
+                    array[sz] = '\\'
+                    array[sz + 1] = marker
+                    sz += 2
+                } else {
+                    val escapedString = ESCAPE_STRINGS[ch]!!
+                    ensureTotalCapacity(sz + escapedString.length)
+                    escapedString.toCharArray(array, sz, 0, escapedString.length)
+                    sz += escapedString.length
+                }
+            } else {
+                array[sz++] = string[i]
+            }
+        }
+        array[sz++] = '"'
+        size = sz
+    }
+
+    actual override fun toString(): String {
+        return String(array, 0, size)
+    }
+
+    private fun ensureAdditionalCapacity(expected: Int) {
+        ensureTotalCapacity(size + expected)
+    }
+
+    private fun ensureTotalCapacity(newSize: Int) {
+        if (array.size <= newSize) {
+            array = array.copyOf(newSize.coerceAtLeast(size * 2))
```

Comment:
Here we can exceed max chars size. If application regularly sends large JSONs, it would still suffer from allocations. Maybe we can make sure that at least one char array always stays in pool? Or recommendation for such applications would be to increase max pool size? Then we should document this.

Reply:
>Maybe we can make sure that at least one char array always stays in pool

It potentially can lead to a tremendous memory footprint of up to 4GB, this is not really desirable.

Even if JSON encoding allocates char arrays for large JSON, it's still okay, what else application author could have expected? We do not guarantee zero allocation anyway. I'd prefer keeping the pool an implementation detail. Ones caring about performance will figure out about pool themselves